src/Microsoft.ML.Tokenizers/Model/SentencePieceBpe.cs

```diff
@@ -0,0 +1,952 @@
+            {
+                for (int i = 0; i < span.Length; i++)
+                {
+                    char c = span[i];
+                    sb.Append(c == LlamaNormalizer.DummyPrefix ? ' ' : c);
+                }
+            }
+
+            static void FlushBytes(ref int bytesCount, ref byte[]? bytesPoolArray, ref char[]? charPoolArray, ref ValueStringBuilder sb)
+            {
+                Debug.Assert(bytesCount >= 1);
+                Debug.Assert(bytesPoolArray is not null);
+
+                int len = Encoding.UTF8.GetMaxCharCount(bytesCount);
+
+                charPoolArray ??= ArrayPool<char>.Shared.Rent(Math.Max(len, 50));
+
+                if (len > charPoolArray.Length)
+                {
+                    Helpers.ArrayPoolGrow(ref charPoolArray, len);
+                }
+
+                int charCount = Helpers.GetChars(bytesPoolArray.AsSpan(0, bytesCount), charPoolArray);
+
+                sb.Append(charPoolArray.AsSpan(0, charCount));
+                bytesCount = -1;
+            }
+
+            static void EncodeByte(int id, int oneByteUtf8EncodingMaxId, int byteCodeToIdOffset, ref int bytesCount, ref byte[]? bytesPoolArray, ref ValueStringBuilder sb)
+            {
+                if (id <= oneByteUtf8EncodingMaxId)
+                {
+                    sb.Append((char)(id - byteCodeToIdOffset));
+                }
+                else
+                {
+                    bytesCount = 1;
+                    bytesPoolArray ??= ArrayPool<byte>.Shared.Rent(50);
+                    bytesPoolArray[0] = (byte)(id - byteCodeToIdOffset);
+                }
+            }
+        }
+
+        // Tries to avoid string allocations if possible.
+        private string GetTokenString(int id, int index, int length, ReadOnlySpan<char> text)
+            => _vocabReverse.TryGetValue(id, out string? token) ? token : text.Slice(index, length).ToString();
+
+        private Dictionary<(int Index, int Len), (int LeftIndex, int LeftLen, int RightIndex, int RightLen)>? Encode(ReadOnlySpan<char> text, BpeSymbol[] symbols)
+        {
+            Debug.Assert(text.Length > 0);
+            Debug.Assert(symbols.Length >= text.Length);
+
+            int symbolIndex = 0;
+            int spanIndex = 0;
+
+            while (spanIndex < text.Length)
+            {
+                int len = (Char.IsHighSurrogate(text[spanIndex]) && spanIndex < text.Length - 1 && Char.IsLowSurrogate(text[spanIndex + 1])) ? 2 : 1;
+
+                BpeSymbol s = new(
+                            prev: symbolIndex == 0 ? -1 : symbolIndex - 1,
+                            next: spanIndex + len >= text.Length ? -1 : symbolIndex + 1,
+                            pieceSpan: (spanIndex, len),
+                            id: UninitializedId,
+                            type: 0);
+
+                symbols[symbolIndex++] = s;
+                spanIndex += len;
+            }
+
+            PriorityQueue<SymbolPair> agenda = new(symbolIndex);
```

Comment:
PriorityQueue doesn't exist in netstandard. Do we have a polyfill for the netstandard build?

Reply:
Yes, we have internal implementation under Util folder. We use it in both ns and .NET too. The reason is it is simpler by using the IComparer in the Queue items instead of using TPriority object every time to enqueue and dequeue.